src/cargo/sources/git/utils.rs

```diff
@@ -893,13 +893,26 @@ fn fetch_with_cli(
         .env_remove("GIT_OBJECT_DIRECTORY")
         .env_remove("GIT_ALTERNATE_OBJECT_DIRECTORIES")
         .cwd(repo.path());
-    config
-        .shell()
-        .verbose(|s| s.status("Running", &cmd.to_string()))?;
-    cmd.exec_with_output()?;
+
+    run_git_cli_with_retries(config, &cmd)?;
+
     Ok(())
 }
 
+/// Run the given command assuming it's spawning `git` CLI process
+/// retrying all the possible transient errors it may fail with.
+fn run_git_cli_with_retries(config: &Config, cmd: &ProcessBuilder) -> CargoResult<()> {
```

Comment:
I don't think we'll have much use for running this in multiple places, so it should be fine to fold this definition into the function above.